src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,693 @@
+      format != "xml-pretty"s) {
+    format = "json-pretty"s;
+  }
+
+  /*
+     match the incoming op-code to one of the registered APIs
+  */
+  auto parsed_cmd = locate_subcmd(match);
+  if (!parsed_cmd.has_value()) {
+    return Maybe_parsed{ std::nullopt };
+  }
+
+  return parsed_command_t{ match,
+                           cmdmap,
+                           format,
+                           parsed_cmd.value(),
+                           parsed_cmd.value()->hook,
+                           full_command_seq };
+}
+
+/*
+  Note: validate_command() is executed with servers_tbl_rwlock held as shared
+*/
+bool AdminSocket::validate_command(const parsed_command_t& parsed,
+                                   const std::string&      command_text,
+                                   ceph::bufferlist&       out) const
+{
+  // did we receive any arguments apart from the command word(s)?
+  if (parsed.m_cmd_seq_len == parsed.m_cmd.length())
+    return true;
+
+  logger().info("{}: validating {} against:{}", __func__, command_text,
+                parsed.m_api->cmddesc);
+
+  stringstream os;  // for possible validation error messages
+  try {
+    // validate_cmd throws on some syntax errors
+    if (validate_cmd(m_cct, parsed.m_api->cmddesc, parsed.m_parameters, os)) {
+      return true;
+    }
+  } catch (std::exception& e) {
+    logger().error("{}: validation failure ({} : {}) {}", __func__,
+                   command_text, parsed.m_cmd, e.what());
+  }
+
+  os << "error: command validation failure ";
+  logger().error("{}: validation failure (incoming:{}) {}", __func__,
+                 command_text, os.str());
+  out.append(os);
+  return false;
+}
+
+seastar::future<> AdminSocket::finalyze_response(
+  seastar::output_stream<char>& out, ceph::bufferlist&& msgs)
+{
+  string outbuf_cont = msgs.to_str();
+  if (outbuf_cont.length() == 0) {
+    outbuf_cont = " {} ";
+  }
+  uint32_t response_length = htonl(outbuf_cont.length());
+  logger().info("asok response length: {}", outbuf_cont.length());
+
+  return out.write((char*)&response_length, sizeof(uint32_t))
+    .then([&out, outbuf_cont]() { return out.write(outbuf_cont.c_str()); });
+}
+
+seastar::future<> AdminSocket::execute_line(std::string cmdline,
+                                            seastar::output_stream<char>& out)
+{
+  return seastar::with_shared(servers_tbl_rwlock, [this, cmdline,
+                                                   &out]() mutable {
+    ceph::bufferlist msgs;
+    Maybe_parsed     oparsed;
+
+    bool ok_to_execute{ false };
+    try {
+      oparsed = parse_cmd(cmdline, msgs);
+      if (oparsed.has_value())
+        ok_to_execute = validate_command(*oparsed, cmdline, msgs);
+    } catch (...) {
+    };
+
+    if (!ok_to_execute) {
+      return finalyze_response(out, std::move(msgs));
+    }
+
+    return seastar::do_with(
+      std::move(*oparsed), std::move(msgs),
+      [this, ok_to_execute, &out](auto&& parsed, auto&& out_buf) {
```

Comment:
Don't think you need to capture ok_to_execute here.

Reply:
Fixed.